spring-cloud-consul-config/src/main/java/org/springframework/cloud/consul/config/watch/ConsulConfigWatch.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2013-2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.consul.config.watch;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.collections4.map.AbstractReferenceMap.ReferenceStrength;
+import org.apache.commons.collections4.map.ReferenceMap;
+import org.springframework.cloud.bootstrap.BootstrapApplicationListener;
+import org.springframework.cloud.consul.config.ConsulConfigProperties;
+import org.springframework.cloud.consul.config.ConsulPropertySource;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.context.EnvironmentAware;
+import org.springframework.core.env.CompositePropertySource;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.Environment;
+import org.springframework.core.env.PropertySource;
+import org.springframework.scheduling.annotation.Scheduled;
+
+import com.ecwid.consul.v1.QueryParams;
+import com.ecwid.consul.v1.Response;
+import com.ecwid.consul.v1.kv.model.GetValue;
+
+import lombok.extern.slf4j.Slf4j;
+
+import static org.springframework.util.Base64Utils.decodeFromString;
+
+/**
+ * A kv store watch that publishes an EnvironmentChangeEvent whenever a value under one of the property sources is changed
+ * 
+ * @author Andrew DePompa
+ */
+@Slf4j
+public class ConsulConfigWatch implements ApplicationEventPublisherAware, EnvironmentAware {
+
+	private ApplicationEventPublisher publisher;
+
+	private ConfigurableEnvironment environment;
+
+	private Set<ConsulPropertySource> consulPropertySources = null;
+
+	private ConsulConfigProperties properties;
+	
+	private Map<String, BigInteger> kvIndexes = new ReferenceMap<String, BigInteger>(ReferenceStrength.SOFT, ReferenceStrength.SOFT);
```

Comment:
What's the reason for using this here? Maybe `ConcurrentReferenceHashMap` from spring-framework could work.